x-pack/plugins/spaces/server/lib/space_schema.ts

```diff
@@ -20,6 +20,17 @@ export const spaceSchema = schema.object({
     },
   }),
   name: schema.string({ minLength: 1 }),
+  solution: offeringBasedSchema({
```

Comment:
I see this schema is currently used in spaces' HTTP API validation and I don't think this will block `solution` from being passed for the `severless` case (`schema.never()`).

It's a bit awkward 😅  bc `offeringBasedSchema` uses `schema.conditional` which relies on this "special" context being passed down that only happens when validating config, not routes:

https://github.com/elastic/kibana/blob/989549c34944fcb459a4eecd04e57aed3c477399/packages/kbn-config/src/config_service.ts#L311-L314

Additionally, `schema.conditional` won't play nice with OAS generation at the moment.

I'd recommend building schema for route APIs dynamically with `if...else`. Something like:

```ts
const isServerless = initializerContext.env.packageInfo.buildFlavor === 'serverless';
...

function registerRoute(deps) {
  const spaceSchema = deps.isServerless ? baseSpaceSchema : baseSpaceSchema.extend({ solution: solutionSchema });
  
  route.put({ ...validate: { body: spaceSchema } })
}
```

Reply:
@jloleysens thanks for bringing it up, I was following the official preferred approach [listed in the doc](https://docs.google.com/document/d/1xNPPnOIvVcya89GeTZByp6YyCWXg1M1v_rrPYT_EBOs/edit#heading=h.z73suhwd57w8), it states the use of `offeringBasedSchema`

> being passed down that only happens when validating config, not routes

checked that once again, indeed it passes it, we have validation on the side of [SpacesClient](https://github.com/elastic/kibana/pull/183986/files#diff-27edbd2fe8e25cb3807f3450e66a0d381fd9d4b62e99d6f88c20edec0d562326R139) though.

I'm okay to switch to dynamic approach.
Do I understand that right, that right now the approaches listed in the doc won't work for API routes at all?

Reply:
> right now the approaches listed in the doc won't work for API routes at all?

`offeringBasedSchema` is not intended to be used on schemas outside of config. That's a failing on our documentation for sure, I'll update the doc and we can add a doc comment to the utility too.